builder/builder_test.go

```diff
@@ -307,3 +339,175 @@ func getAppHashFromEVM(ethState *state.StateDB, header *monomer.Header) (common.
+		t,
+		cosmAddr.String(),
+		common.HexToAddress("0x12345abcde").String(),
+		math.NewIntFromBigInt(depositTxETH.Value()),
+	)
+
+	b := createBuilder(pool, app, blockStore, txStore, eventBus, g.ChainID, ethstatedb)
+
+	// Prepare the payload
+	txs := bfttypes.Txs{depositTxs, withdrawalTx}
+	payload := &builder.Payload{
+		InjectedTransactions: txs,
+		GasLimit:             1000000000000,
+		Timestamp:            g.Time + 1,
+		NoTxPool:             true,
+	}
+
+	// Build a block with the payload
+	builtBlock, preBuildInfo, postBuildInfo := buildBlock(t, b, app, payload)
+
+	// Test deposit was received
+	expectedDepositAmount := fmt.Sprintf("%sETH", depositTxETH.Value().String())
+	checkDepositTXResult(t, txStore, depositTxs, expectedDepositAmount, cosmAddr.String())
+
+	withdrawalTxResult, err := txStore.Get(bfttypes.Tx(withdrawalTx).Hash())
+	require.NoError(t, err)
+	require.NotNil(t, withdrawalTxResult)
+	require.Truef(t, withdrawalTxResult.Result.IsOK(), "Expected the withdrawal transaction to be successful, but it failed")
+
+	// Verify block creation
+	genesisBlock, err := blockStore.BlockByHeight(uint64(preBuildInfo.GetLastBlockHeight()))
+	require.NoError(t, err)
+	gotBlock, err := blockStore.HeadBlock()
+	require.NoError(t, err)
+
+	ethStateRoot := gotBlock.Header.StateRoot
+	header := monomer.Header{
+		ChainID:    g.ChainID,
+		Height:     uint64(postBuildInfo.GetLastBlockHeight()),
+		Time:       payload.Timestamp,
+		ParentHash: genesisBlock.Header.Hash,
+		StateRoot:  ethStateRoot,
+		GasLimit:   payload.GasLimit,
+	}
+	wantBlock, err := monomer.MakeBlock(&header, txs)
+	require.NoError(t, err)
+	verifyBlockContent(t, wantBlock, gotBlock, builtBlock)
+
+	// We expect two transactions: one combined transaction (l1InfoTx + depositTxs) and one withdrawal transaction (withdrawalTx).
+	require.Equal(t, 2, builtBlock.Txs.Len(), "Expected the built block to contain 2 transactions: depositTxs and withdrawalTx")
+	require.Equal(t, 2, gotBlock.Txs.Len(), "Expected the built block to contain 2 transactions: depositTxs and withdrawalTx")
+
+	// Test parseWithdrawalMessages
+	testParseWithdrawalMessages(t, withdrawalTxResult)
+
+	expectedStateRoot := wantBlock.Header.StateRoot
+	gotStateRoot := gotBlock.Header.StateRoot
+	require.Equal(t, expectedStateRoot, gotStateRoot, "Expected the built block to contain state root hash")
+
+	for i := range gotBlock.Txs {
+		select {
+		case event, ok := <-subscription.Out():
+			if !ok {
+				require.FailNow(t, "event channel closed unexpectedly")
+			}
+			data := event.Data()
+			require.IsType(t, bfttypes.EventDataTx{}, data)
+			e, ok := data.(bfttypes.EventDataTx)
+			require.True(t, ok)
+			// Deposit transaction
+			if i == 0 {
+				require.Equal(t, e.GetResult().Events[0].Attributes[0].Value, "/rollup.v1.MsgApplyL1Txs")
+			}
+			// Withdrawal transaction
+			if i == 1 {
+				require.Equal(t, e.GetResult().Events[0].Attributes[0].Value, "/rollup.v1.MsgInitiateWithdrawal")
+			}
```

Comment:
Instead of using indexes to differentiate between the deposit and withdrawal txs, could you check the event type instead to differentiate? That seems cleaner to me and will help with readability

Reply:
@natebeauregard Their types are both "message"

```
{message [{action /rollup.v1.MsgApplyL1Txs true} {sender cosmos1v9jxguspv4h2u true} {msg_index 0 true}]}

{message [{action /rollup.v1.MsgInitiateWithdrawal true} {sender cosmos1wmufz6m7jjekmk7kafqvpeaveg4gx993u9qvsf true} {msg_index 0 true}]}
```

I'll go ahead and use the type instead of index. Thanks

Reply:
@natebeauregard on second glance, this will cause an error since both if statements will hit on first iteration

Reply:
What if you instead just extracted out the `select` statement into a helper func as suggested in my comment above? Then you should be able to loop over the deposit/withdrawal expected event attributes without needing to rely on indexing? I feel like that would be cleaner than using indexes to verify the expected events